app/admin/resources/resource_factory.ts

```diff
@@ -0,0 +1,252 @@
+  RecordActionResponse,
+  RecordJSON,
+  ResourceOptions,
+  ResourceWithOptions,
+} from "adminjs";
+
+import { MultipartFile } from "@adonisjs/core/types/bodyparser";
+import { LucidModel } from "@adonisjs/lucid/types/model";
+
+import FilesService from "#services/files_service";
+
+export interface ResourceNavigation {
+  name: string;
+  icon: string;
+}
+
+export interface ResourceBuilder {
+  navigation: ResourceNavigation;
+  builders: ResourceInfo[];
+}
+
+type ImprovedRecordJSON = Omit<RecordJSON, "params"> & {
+  params: Record<string, ParamsTypeValue>;
+};
+type ImprovedRecordActionResponse = Omit<RecordActionResponse, "record"> & {
+  record: ImprovedRecordJSON;
+};
+export type ActionMap = Record<string, Partial<Action<ActionResponse>>>;
+type SubAction =
+  | Partial<Action<ActionResponse>>
+  | Partial<Action<RecordActionResponse>>;
+type HookReturnValue = Partial<Action<RecordActionResponse>>;
+
+export interface ResourceInfo {
+  forModel: LucidModel;
+  additionalProperties?: Record<string, PropertyOptions>;
+  additionalActions?: ActionMap;
+  additionalOptions?: ResourceOptions;
+  addImageHandling?: boolean;
+}
+
+const hideOnEdit = {
+  edit: false,
+  show: true,
+  new: false,
+  list: true,
+};
+
+const hideOnShow = {
+  list: false,
+  show: false,
+  new: true,
+  edit: true,
+};
+
+const DB_DRIVER = "postgres";
+
+const readOnlyTimestamps = {
+  createdAt: {
+    type: "datetime" as const,
+    isVisible: {
+      edit: false,
+      show: true,
+      list: false,
+      filter: true,
+    },
+  },
+  updatedAt: {
+    type: "datetime" as const,
+    isVisible: {
+      edit: false,
+      show: true,
+      list: false,
+      filter: true,
+    },
+  },
+} as const;
+
+export class ResourceFactory {
+  private registeredResources: ResourceBuilder[];
+
+  constructor(registeredResources?: ResourceBuilder[]) {
+    this.registeredResources = registeredResources ?? [];
+  }
+
+  public registerResource(resourceBuilder: ResourceBuilder) {
+    this.registeredResources.push(resourceBuilder);
+  }
+
+  public buildResources() {
+    return this.registeredResources
+      .map((resourceBuilder) => this.createResources(resourceBuilder))
+      .flat(1);
+  }
+
+  public createResources(
+    resourceBuilder: ResourceBuilder,
+  ): ResourceWithOptions[] {
+    return resourceBuilder.builders.map((resourceInfo) => {
+      return ResourceFactory.createResource(
+        resourceInfo.forModel,
+        resourceBuilder.navigation,
+        resourceInfo.additionalProperties,
+        resourceInfo.additionalActions,
+        resourceInfo.additionalOptions,
+        resourceInfo.addImageHandling,
+      );
+    });
+  }
+
+  private static createResource(
+    resourceModel: LucidModel,
+    resourceNavigation: ResourceNavigation,
+    additionalProperties?: Record<string, PropertyOptions>,
+    additionalActions?: ActionMap,
+    additionalOptions?: ResourceOptions,
+    addImageHandling = false,
+  ): ResourceWithOptions {
```

Comment:
with so many parameters (some optional) I'd consider passing everything in an object to immitate keyword arguments and make it easier to read what's going on at the call site (plus you might be able to use `...` to merge both objects)